### PR TITLE
upcoming: [M3-8020] – Add "Disk Encryption" section to Linode Rebuild modal

### PR DESCRIPTION
**Pull request by dwiley-akamai**
## Description 📝
Add "Disk Encryption" section to Linode Rebuild modal.

> [!NOTE]
> I came across an existing edge case where it is possible that a field absent from the Rebuild dialog can still cause the Rebuild request to fail silently. See M3-8237.

## Changes  🔄
- Add `isLKELinode`, `linodeIsInDistributedRegion`, and `isInRebuildFlow` props to `AccessPanel.tsx`
- Several props added to `RebuildFromImage` & `RebuildFromStackScript`

## Target release date 🗓️
6/24

## Preview 📷
<details>
<summary>Screenshots</summary>

**Standard Linode in a region that supports LDE**:

Screenshot 2024-06-13 at 11 50 23 AM

**Linode in Distributed region**:

Screenshot 2024-06-13 at 11 27 05 AM

**LKE Linode**:
Screenshot 2024-06-13 at 11 49 26 AM

</details>

### Prerequisites
1. Have the `linode_disk_encryption` tag on your account
2. Point at the alpha/dev environment
3. Have the Linode Disk Encryption (LDE) flag in our dev tool toggled on

### Verification steps
Without the tag and/or with the LDE flag in our dev tool toggled off, you should not see any LDE-related things in the UI. Otherwise,

- for a non-LKE linode in Newark, the "Encrypt Disk" checkbox in the Rebuild modal should be enabled and you should be able to successfully rebuild it
- for an LKE linode, the "Encrypt Disk" checkbox in the Rebuild modal should be disabled with:
  - section description copy reading: `Secure this Linode using data at rest encryption. The disk encryption setting for Linodes added to a node pool will not be changed after rebuild.`
  - tooltip text reading: `The Encrypt Disk setting cannot be changed for a Linode attached to a node pool.`
- for a linode in a Distributed region,  the "Encrypt Disk" checkbox in the Rebuild modal should be disabled with:
  - section description copy reading: `Distributed Compute Instances are secured using disk encryption.`
  -  tooltip text reading: `The Encrypt Disk setting cannot be changed for distributed instances.`

The last two cases can be easily tested by adjusting this code block as needed and turning the MSW on:

https://github.com/linode/manager/blob/0e2d367d192f9a28b92b5b37ebd7f566c345ef5c/packages/manager/src/mocks/serverHandlers.ts#L768-L778

```typescript
      linodeFactory.build({
        backups: { enabled: false },
        id,
        label: 'Gecko Edge Test',
        lke_cluster_id: null, // change to a number to see the LKE Linode state
      })
```

Things should work when rebuilding from an image, a Community StackScript, and an Account StackScript.

Rebuilding an encrypted linode --> the "Encrypt Disk" checkbox should be checked already in the modal to reflect the current configuration

There should be no adverse effects observed if you switch to the prod environment and turn the feature flag off.

## As an Author I have considered 🤔
- [X] 👀 Doing a self review
- [X] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [X] 🤏 Splitting feature into small PRs
- [X] ➕ Adding a changeset
- [X] 🧪 Providing/Improving test coverage
- [X] 🔐 Removing all sensitive information from the code and PR description
- [X] 🚩 Using a feature flag to protect the release
- [X] 👣 Providing comprehensive reproduction steps
- [X] 📑 Providing or updating our documentation
- [X] 🕛 Scheduling a pair reviewing session
- [X] 📱 Providing mobile support
- [X] ♿  Providing accessibility support

### CODE REVIEW

File: packages/manager/src/features/Events/factories/tax.tsx
Comment (dwiley-akamai) [inline packages/manager/src/features/Events/factories/tax.tsx:1]: Diff generated by linter

File: packages/manager/src/components/AccessPanel/AccessPanel.tsx [hunk 1]
```diff
@@ -97,6 +104,30 @@ export const AccessPanel = (props: Props) => {
...
+    if (linodeIsInDistributedRegion) {
+      return ENCRYPT_DISK_DISABLED_REBUILD_DISTRIBUTED_REGION_REASON;
+    }
+
+    if (!regionSupportsDiskEncryption) {
+      return DISK_ENCRYPTION_UNAVAILABLE_IN_REGION_COPY;
+    }
+
+    return '';
+  };
```
Comment (dwiley-akamai) [inline packages/manager/src/components/AccessPanel/AccessPanel.tsx:129]: I preferred this over the initial commit's nested ternaries

<details>

```typescript

const encryptDiskDisabledReason = isLKELinode
    ? ENCRYPT_DISK_DISABLED_REBUILD_LKE_REASON
    : linodeIsInDistributedRegion
    ? ENCRYPT_DISK_DISABLED_REBUILD_DISTRIBUTED_REGION_REASON
    : !regionSupportsDiskEncryption
    ? DISK_ENCRYPTION_UNAVAILABLE_IN_REGION_COPY
    : '';

```
</details>

**Review by abailly-akamai — APPROVED** [review]
Nice PR! well tested and documented

Was able to confirm:
- for a non-LKE linode in Newark, the "Encrypt Disk" checkbox in the Rebuild modal should be enabled and you should be able to successfully rebuild it ✅ 
- for an LKE linode, the "Encrypt Disk" checkbox in the Rebuild modal should be disabled with tooltip text reading: The Encrypt Disk setting cannot be changed for a Linode attached to a node pool. ✅ 
- for a linode in a Distributed region, the "Encrypt Disk" checkbox in the Rebuild modal should be disabled with tooltip text reading: The Encrypt Disk setting cannot be changed for distributed instances. ✅ 

and confirmed it is also still disabled in unsupported regions (ex: London)

File: packages/manager/src/features/Linodes/LinodesDetail/LinodeRebuild/LinodeRebuildDialog.tsx [hunk 2]
```diff
@@ -51,11 +57,24 @@ export const LinodeRebuildDialog = (props: Props) => {
   const unauthorized = isReadOnly;
   const disabled = hostMaintenance || unauthorized;
 
+  // LDE-related checks
+  const isEncrypted = linode?.disk_encryption === 'enabled' ? true : false;
```
[the remark this answers is not in the document]
Comment (dwiley-akamai) [reply packages/manager/src/features/Linodes/LinodesDetail/LinodeRebuild/LinodeRebuildDialog.tsx]: Thanks for bringing this to my attention, upon second look we should be fine with just the equality check

**Comment by carrillo-erik** [comment]
Good job catching this edge case. I was able to follow the Verification Steps and did not notice find any issues with functionality or the UI. Approved ✅.

**Review by carrillo-erik — APPROVED** [review]

**Pull request merged**